### PR TITLE
function for updating the existing role instead of removing, then adding one

### PR DESCRIPTION
**Pull request by peterwilsoncc**
https://core.trac.wordpress.org/ticket/54572

### CODE REVIEW

File: src/wp-includes/capabilities.php [hunk 1]
```diff
@@ -849,6 +849,31 @@ function add_role( $role, $display_name, $capabilities = array() ) {
...
+/**
+ * Update an existing role. Create a new role if it doesn't exist.
+ *
+ * Modify the display name and/or capabilities for an existing role. If the
+ * role doesn't exist then a new role is created.
+ *
+ * The capabilities are defined in the following format `array( 'read' => true );`
+ * To explicitly deny a role a capability you set the value for that capability to false.
+ *
+ * @param string      $role         Role name.
```
Comment (peterwilsoncc) [inline src/wp-includes/capabilities.php:1044]: `@since` x2, just waiting to confirm.

File: tests/phpunit/tests/user/capabilities.php [hunk 2]
```diff
@@ -980,6 +980,461 @@ public function test_add_role() {
...
+		add_role( $role_name, 'Janitor', $expected_caps );
+		$this->flush_roles();
+
+		$this->assertTrue(
+			$wp_roles->is_role( $role_name ),
+			"The $role_name role was not created"
+		);
+
+		$this->assertNotInstanceOf(
+			'WP_Roles',
```
Comment (costdev) [inline tests/phpunit/tests/user/capabilities.php]: I think this is a typo and it should be `WP_Role`.

Pinging @peterwilsoncc for a confidence check.

File: src/wp-includes/capabilities.php [hunk 3]
```diff
@@ -1029,6 +1029,28 @@ function add_role( $role, $display_name, $capabilities = array() ) {
...
 
+/**
+ * Updates an existing role. Creates a new role if it doesn't exist.
+ *
+ * Modifies the display name and/or capabilities for an existing role.
+ * If the role does not exist then a new role is created.
+ *
+ * The capabilities are defined in the following format: `array( 'read' => true )`.
+ * To explicitly deny a role a capability, set the value for that capability to false.
+ *
```
Comment (costdev) [inline src/wp-includes/capabilities.php:1043]:
```suggestion
 *
 * @since 6.1.0
 *
```

File: src/wp-includes/class-wp-roles.php [hunk 4]
```diff
@@ -170,6 +170,74 @@ public function add_role( $role, $display_name, $capabilities = array() ) {
...
 
+	/**
+	 * Updates an existing role. Creates a new role if it doesn't exist.
+	 *
+	 * Modifies the display name and/or capabilities for an existing role.
+	 * If the role does not exist then a new role is created.
+	 *
+	 * The capabilities are defined in the following format: `array( 'read' => true )`.
+	 * To explicitly deny a role a capability, set the value for that capability to false.
+	 *
```
Comment (costdev) [inline src/wp-includes/class-wp-roles.php:184]:
```suggestion
	 *
	 * @since 6.1.0
	 *
```

**Review by costdev — CHANGES_REQUESTED** [review]

File: tests/phpunit/tests/user/capabilities.php (hunk 2 above)
Comment (peterwilsoncc) [reply tests/phpunit/tests/user/capabilities.php]: Yes, it should be, however changing it to `WP_Role` also causes all of these tests to fail.

I need to check out why.

File: tests/phpunit/tests/user/capabilities.php [hunk 5]
```diff
@@ -980,6 +980,461 @@ public function test_add_role() {
...
+			update_role( $role_name, 'Janitor', $expected_caps ),
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/user/capabilities.php]:
```php
			update_role( $role_name, 'Janitor', $expected_caps ),
```

This is why, it's using the incorrect variable names -- it should be using the ones passed by the data provider.

**Comment by peterwilsoncc** [comment]
I've pushed a few changes:

* https://github.com/WordPress/wordpress-develop/pull/2585/commits/9e9349cf8a16ae02098bfe63d9ec4d6f9f2a3cda: Since tags
* https://github.com/WordPress/wordpress-develop/pull/2585/commits/804e988ee8e5e5a4ac5ee5dfc834dfd4efa8859c Fixed the broken test 
* https://github.com/WordPress/wordpress-develop/pull/2585/commits/f13d2e5e1a0de6a9011416032ffd6b989e59dfcf renamed unhappy path test to follow the convention of using data providers with a data_ prefix replacing test_

File: src/wp-includes/class-wp-roles.php [hunk 6]
```diff
@@ -138,20 +138,20 @@ public function reinit() {
 	}
 
 	/**
-	 * Add role name with capabilities to list.
+	 * Adds a role name with capabilities to the list.
```
Comment (costdev) [inline src/wp-includes/class-wp-roles.php]: Should the docs changes for `add_role()` be included in this PR? Or will they be cherrypicked and committed separately?

Comment (SergeyBiryukov) [reply src/wp-includes/class-wp-roles.php]: Good call, I've committed these changes separately in https://core.trac.wordpress.org/changeset/53436.

**Comment by dream-encode** [comment]
Merged into core in https://core.trac.wordpress.org/changeset/54213.

**Pull request closed without merging**